serverside-javascript-best-practices.md

```diff
@@ -320,6 +350,53 @@ createBooking.call(self, product, params, 'upgrades', 'supplier', 'payment', fun
 
 
 
+## Handling heavily sequential async code with promises
+
+* !! Using promises for heavily sequential async code has the downside of hiding the data being passed between functions. We lose the ability to wrap each layer in a closure to selectively pass the correct data betweeen the correct entites. Eveything must flow through the middle to get to the end. This breaks our rules on single-responsibility, but allows for slightly cleaner orchestration:
+* !! Sinon promises do not support sandboxing yet, so you'll have to manually restore everything by hand
+* !! Its impossible to prove that the data flow between promises is correct, only that data moves between them.
+
+```javascript
+createBooking._orchestrate = function(data) {
+  return createBooking._sendBook1Request(data)
+    .then(createBooking._generatePaymentRequest)
+    .then(createBooking._processPayment)
+    .then(createBooking._sendBook2Request)
+    .then(createBooking._generateBookingReference);
+};
+```
+```javascript
+require('sinon-bluebird');
```

Comment:
[sinon-as-promised](https://www.npmjs.com/package/sinon-as-promised) might be the best option here since it works with different promises implementations as long the implementation provides a constructor which should cover ES6 promises and the most popular promises libraries (Bluebird, Q, ...).